designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+
+```js
+const standardConfig = require("eslint-config-standard");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+exports.config = myConfig;
+```
+
+Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)
+
+Similarly, the `Config.assign()` method lets users modify their config by merging in changes from other configs, such as:
+
+```js
+const standardConfig = require("eslint-config-standard");
+const personalConfig = require("@personal/eslint-config");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+Config.assign(myConfig, personalConfig);
+
+exports.config = myConfig;
+```
+
+This code example merges `personalConfig` into `myConfig` using the same functionality that ESLint currently uses to merge configs.
+
+#### The `RuleLoader` Class
+
+The purpose of the `RuleLoader` class is aid in loading rules to replace `--rulesdir` functionality and has this form:
+
+```js
+class RuleLoader {
+
+    // create a new instance
+    constructor() {}
+
+    // mimic `--rulesdir` loading
+    loadRulesFromDirectory(directory) {}
+}
+```
+
+In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `ruledefs` and then specify the rules from a directory, such as:
+
+```js
+const { Config, RuleLoader } = require("@eslint/config");
+const ruleLoader = new RuleLoader();
+
+exports.config = new Config({
+    ruledefs: {
+        custom: ruleLoader.loadFromDirectory("./custom-rules")
+    },
+    rules: {
+        "custom/my-rule": "error"
+    }
+});
+```
+
+The `RuleLoader#loadFromDirectory()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).
+
+### Advanced Configs
+
+Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:
+
+```js
+exports.config = (context) => {
+
+    // do something
+
+    return {
+        rules: {
+            "semi": ["error", "always"]
+        }
+    };
+
+};
+```
+
+The `context` object has the following members:
+
+* `core` - information about the ESLint core that is using the config
```

Comment:
I was purposely trying not to use "eslint" for this key in the off case that a config ends up in some ESLint wrapper like Standard or XO and to make it clear that this info was coming from the core that was running the config rather than the application itself (hence why you can't get plugin info, for example).

I'm not tied to `core`, though `eslintInfo` seems a bit verbose to me.

Reply:
If a wrapper like Standard is running a config, is the expectation that the version property would represent the version of ESLint, or the version of Standard?

I had expected that it would represent the version of ESLint. For example, suppose ESLint v6.2.0 adds a particular rule option, and my config checks the `version` property to determine whether it should enable that option (based on whether the version is at least v6.2.0). If the `standard@4.1.0` package happens to depend on `eslint@6.3.0`, then passing `v4.1.0` as the `core.version` property doesn't seem like it would make much sense, because the version of Standard being used is largely irrelevant to what ESLint features are available.

As a result, including `eslint` in the key seems like it would make sense to me (or your aversion to including `eslint` in the key makes me suspect I'm misunderstanding something).

Reply:
The version would represent the ESLint core being used.

My idea here is that other tools might want to white label ESLint, even changing the name of the config file, so `eslint` might not necessarily mean anything to those users. They might also want to, for example, have other core rules (so `hasRule()`) could be different for a white label.

It's possible I'm overthinking it, though. I found naming this thing difficult. I wanted a way to say "this ESLint core has this version and contains these rules." I'm open to another name, but I'm not a fan of `eslintInfo`.

Reply:
Thanks for clarifying. Given this reasoning I'm fine with keeping the name `core`.